src/ci_ex_states/excited_state_solver.h

```diff
@@ -0,0 +1,148 @@
+/*
+ * @BEGIN LICENSE
+ *
+ * Forte: an open-source plugin to Psi4 (https://github.com/psi4/psi4)
+ * that implements a variety of quantum chemistry methods for strongly
+ * correlated electrons.
+ *
+ * Copyright (c) 2012-2019 by its authors (see COPYING, COPYING.LESSER, AUTHORS).
+ *
+ * The copyrights for code used from other parties are included in
+ * the corresponding files.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ *
+ * @END LICENSE
+ */
+
+#ifndef _excited_state_solver_h_
+#define _excited_state_solver_h_
+
+#include "sparse_ci/determinant_hashvector.h"
+#include "base_classes/active_space_method.h"
+#include "base_classes/rdms.h"
+#include "sparse_ci/sparse_ci_solver.h"
+
+#ifdef _OPENMP
+#include <omp.h>
+#else
+#define omp_get_max_threads() 1
+#define omp_get_thread_num() 0
+#define omp_get_num_threads() 1
+#endif
+
+namespace forte {
+class SelectedCIMethod;
+
+class ExcitedStateSolver : public ActiveSpaceMethod {
```

Comment:
@tyzhang1993: I am not sure this class hierarchy is consistent with how we designed the new base classes. Other methods implement excited states (e.g. FCI) without creating an excited state class. Can we please discuss a different design where the various algorithms are implemented in a base class of SelectedCI and methods like ACI are derived directly from SelectedCI?

Reply:
Yes, it is a good point.
I had this design in order to have functionality really separated. Unlike FCI where the excited states algorithm is straightforward, the excited states algorithm for selected CI methods is pretty independent and has many variations.
The role of excited state solver is to determine how to compute a state, it calls the selected CI method.
The role of selected CI method is performing a selection of important determinants and optimize the coefficients.
Excited state solver drives the computation of selected CI wave function. If they are in one class, I think it is strange to have a class drive itself.
For example, in the original ACI implementation, in order to "drive ACI" performing root_orthogonalized computation, the class variable nroot_ has to be set to 1 before solving the PQ space, then set back to original value. https://github.com/evangelistalab/forte/blob/e82b201ecd70cacb81fd0b7445cec6348b07ba8d/src/sci/aci.cc#L1663-L1672
In my opinion, different excited state algorithms are actually different subclasses of ExcitedStateSolver, in commit https://github.com/evangelistalab/forte/pull/116/commits/600b6544911dfaa7bee141293ec67b28200d2b84 , there are not only the superclass ExcitedStateSolver, but also GroundStateSolver, RootOrthoSolver etc. These excited states algorithms are very different from each other, having derived classes can make code more clear and avoid if statements. It is like our having different variations of sigmavector classes.
https://github.com/evangelistalab/forte/blob/9ee177bff82d50b6e1e1de9487f3151598afcc9b/src/ci_ex_states/excited_state_solver.cc#L169-L188
I didn't separate ExcitedStateSolver into these subclasses in this PR because I do not want the PR to change code too dramatic. In this PR I just move code around to create the structure. but I believe put excited state algorithms into subclasses of  ExcitedStateSolver should be done in the following PRs.
Because the ExcitedStateSolver and SelectedCIMethod are independent of each other and they both have many different variations, I think they should be different classes.